Apps/Encounter/src/Delegates/RestMSPVisitDelegate.cs

```diff
@@ -0,0 +1,161 @@
+    using Microsoft.Extensions.Configuration;
+    using Microsoft.Extensions.Logging;
+
+    /// <summary>
+    /// ODR Implementation for Rest Medication Statements.
+    /// </summary>
+    public class RestMSPVisitDelegate : IMSPVisitDelegate
+    {
+        private const string ODRConfigSectionKey = "ODR";
+
+        private readonly ILogger logger;
+        private readonly ITraceService traceService;
+        private readonly IHttpClientService httpClientService;
+        private readonly IConfiguration configuration;
+        private readonly ODRConfig odrConfig;
+        private readonly Uri baseURL;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="RestMSPVisitDelegate"/> class.
+        /// </summary>
+        /// <param name="logger">Injected Logger Provider.</param>
+        /// <param name="traceService">Injected TraceService Provider.</param>
+        /// <param name="httpClientService">The injected http client service.</param>
+        /// <param name="configuration">The injected configuration provider.</param>
+        public RestMSPVisitDelegate(
+            ILogger<RestMSPVisitDelegate> logger,
+            ITraceService traceService,
+            IHttpClientService httpClientService,
+            IConfiguration configuration)
+        {
+            this.logger = logger;
+            this.traceService = traceService;
+            this.httpClientService = httpClientService;
+            this.configuration = configuration;
+            this.odrConfig = new ODRConfig();
+            this.configuration.Bind(ODRConfigSectionKey, this.odrConfig);
+            if (this.odrConfig.DynamicServiceLookup)
+            {
+                string? serviceHost = Environment.GetEnvironmentVariable($"{this.odrConfig.ServiceName}{this.odrConfig.ServiceHostSuffix}");
+                string? servicePort = Environment.GetEnvironmentVariable($"{this.odrConfig.ServiceName}{this.odrConfig.ServicePortSuffix}");
+                Dictionary<string, string> replacementData = new Dictionary<string, string>()
+                {
+                    { "serviceHost", serviceHost! },
+                    { "servicePort", servicePort! },
+                };
+                this.baseURL = new Uri(StringManipulator.Replace(this.odrConfig.BaseEndpoint, replacementData) !);
+            }
+            else
+            {
+                this.baseURL = new Uri(this.odrConfig.BaseEndpoint);
+            }
+
+            logger.LogInformation($"ODR Proxy URL resolved as {this.baseURL.ToString()}");
+        }
+
+        /// <inheritdoc/>
+        public async Task<RequestResult<MSPVisitHistoryResponse>> GetMSPVisitHistoryAsync(ODRHistoryQuery query, string hdid, string ipAddress)
+        {
+            using ITracer tracer = this.traceService.TraceMethod(this.GetType().Name);
+            if (query == null)
+            {
+                throw new ArgumentNullException(nameof(query), "Query cannot be null");
```

Comment:
are these necessary? I think we have no nullable types to prevent this

Reply:
right, also these checks should be done at the caller level (i.e. Service or Controller) instead of here. It will allow us to follow the Template Method design pattern where we (as dev) can focus only on the implementation of Delegate instead of validations/checks.

Reply:
removed